catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/ResultIterable.java

```diff
@@ -201,19 +210,50 @@ public Result next() {
       return results.next();
     }
 
+    @SuppressWarnings("squid:CommentedOutCodeLine")
     private void fetchNextResults() {
       queryCopy.setStartIndex(currentIndex);
 
       try {
         SourceResponse response = queryFunction.query(queryRequestCopy);
 
-        if (response.getResults().size() == 0) {
+        final List<Result> resultList = response.getResults();
+
+        // Because some of the results may be filtered out by the catalog framework's
+        // plugins, we need a way to know the actual page size and increment currentIndex based
+        // on that number instead of using the result list size.
+        // If the property is not present, we will have no option but to fallback to the size
+        // of the (potentially filtered) resultList.
+        //
+        // This means that if the filtered results size is zero, but the raw number of results
+        // had been greater than zero, we will not find results beyond the filtered gap. In practice
+        // this should not happen, as queries will run through the QueryOperations.query() method;
+        // however, should a user ever construct a QueryFunction that does NOT rely on that method,
+        // there is no guarantee that this property will be properly set.
+        int actualResultSize =
+            (int)
+                Optional.ofNullable(response.getProperties())
+                    .map(m -> m.get("actualResultSize"))
+                    .orElse(resultList.size());
+
+        if (actualResultSize == 0) {
           finished = true;
           return;
         }
+        currentIndex += actualResultSize;
+
+        final List<Result> dedupedResults =
```

Comment:
I try to not mutate collections returned from api's since it can produce some nasty issues, so I will pre create the list the same size as the results